[FeatureHighlight] View should not be loaded unless it's explicitly called or presented.

mohammadcazig opened the pull request:
close #4232

mohammadcazig committed:
Feature Highlight must not load the view unless the view property is called or it's presented.

jverkoey commented:
Please add a description of what this bug is fixing / what the original problem was.

jverkoey commented on components/FeatureHighlight/src/MDCFeatureHighlightViewController.m:
```diff
@@ -57,6 +59,38 @@ - (nonnull instancetype)initWithHighlightedView:(nonnull UIView *)highlightedVie
+  [self setupFeatureHighlightView];
+}
+
+- (void)setupFeatureHighlightView {
```
Do you expect someone to call this method other than viewDidLoad? If not, please keep the logic in viewDidLoad.

jverkoey commented on components/FeatureHighlight/src/MDCFeatureHighlightViewController.m:
```diff
@@ -232,6 +233,20 @@ - (void)dismiss:(BOOL)accepted {
+- (void)applyMDCFeatureHighlightViewDefaults {
+  _outerHighlightColor = [self MDCFeatureHighlightDefaultOuterHighlightColor];
+  _innerHighlightColor = [self MDCFeatureHighlightDefaultInnerHighlightColor];
+}
```
Similar to the viewDidLoad logic above, please consider moving this into the init method.

jverkoey commented on components/FeatureHighlight/src/MDCFeatureHighlightViewController.m:
```diff
@@ -232,6 +233,20 @@ - (void)dismiss:(BOOL)accepted {
+}
+
+- (UIColor *)MDCFeatureHighlightDefaultOuterHighlightColor {
+  return [[UIColor blueColor] colorWithAlphaComponent:kMDCFeatureHighlightOuterHighlightAlpha];
```
How do you expect these methods to be used? If they are simply to initialize the ivars, please consider removing the method indirection altogether and simply assign the ivar in the init method.

jverkoey reviewed (dismissed):
I'm seeing a common pattern in this PR where code is being pulled out of init/viewDidLoad methods into helper methods. This is not typically a pattern that we employ because doing so encourages the ability to call code paths that should only ever happen in the init or viewDidLoad methods.

I'm also seeing single-line methods that are being created that only have one reference in the code — please be cautious with regard to over-factoring the code.

mohammadcazig committed:
fixed the tests,
added new tests for loading view,
Moved the logic for accessibility hint default to view controller,
Match the pattern of of MDCAlertController for not loading the view.

jverkoey commented on components/FeatureHighlight/src/MDCFeatureHighlightViewController.m:
```diff
@@ -17,28 +17,31 @@
 @interface MDCFeatureHighlightViewController () <UIViewControllerTransitioningDelegate>
-@property(nonatomic, strong) MDCFeatureHighlightView *view;
+@property(nonatomic, nullable, weak) MDCFeatureHighlightView *featureHighlightView;
 @end
```
Why is this weak now?

ianegordon replied:
Originally we had a strong reference to MDCCustomView that was aliasing the standard UIView *view.
This can cause confusion for developers as we are invoking methods and properties on our customView that don't exist in UIView.
To put it another way:
How are we setting self.view.titleFont = XYZ when UIView doesn't have a titleFont property?

jverkoey replied:
I'm not so concerned about the type of the object as much as why it's `weak`. Why does it have to be `weak`?

jverkoey replied:
E.g. UIViewController's `.view` type is:

```objc
@property(null_resettable, nonatomic, strong) UIView *view;
```

jverkoey commented on components/FeatureHighlight/src/MDCFeatureHighlightViewController.m:
```diff
@@ -152,61 +171,55 @@ - (void)viewWillTransitionToSize:(CGSize)size
-- (UIFont *)bodyFont {
-  return self.view.bodyFont;
+  _bodyFont = bodyFont;
+  if (self.featureHighlightView) {
```
Here and throughout you don’t need the if checks because of nil handling.

ianegordon replied:
I explicitly asked @mohammadcazig to add the checks to explicitly indicate that the view may not exist.  In the past we haven't always been properly observing / maintaining a view controller's view lifecycle.

jverkoey replied:
That's a fair intent, but this pattern is unusual for Objective-C.

Did you mean to ask for an `isViewLoaded` check?

ianegordon replied:
If we call the self.view getter, loadView will get invoked.  I recommend this pattern as an explicit indicator to developers that the view may not exist yet. We haven't always been as diligent as we should on view life cycles.

jverkoey replied:
`self.featureHighlightView` is exactly the pattern you're referring to here - we're accessing a view on the view controller, where typically accessing the view has the side-effect of loading it. This pattern as-is is not strictly accomplishing anything (because if the view did lazily load, the if check would load it) and if it doesn't load, then the nil behavior of Objective-C would make the contained logic a no-op.

We either

1. Use `isViewLoaded` to avoid running any logic that should only happen once the main view has loaded.
2. We use standard Objective-C notation for setting values on properties that might be nil (i.e. we don't do `if (object) object.value = something;`).

ianegordon replied:
I agree isViewLoaded is the most obvious / explicit test over self.customView != nil.
@mohammadcazig does that seem like a reasonable approach?

mohammadcazig replied:
so basically we go back to defining view as MDCFeatureHighlightViewController and use self.view to access featurehighlight view correct?

mohammadcazig replied:
@ianegordon @jverkoey

jverkoey replied:
I think that for the sake of this PR it would be most accurate to replace `if (self.featureHighlightView)` with `if (self.isViewLoaded)` calls.

ianegordon replied:
+1 to @jverkoey suggestion.

ianegordon commented on components/FeatureHighlight/src/MDCFeatureHighlightViewController.m:
```diff
@@ -48,15 +51,48 @@ - (nonnull instancetype)initWithHighlightedView:(nonnull UIView *)highlightedVie
+
+  self.featureHighlightView.displayedView = _displayedView;
+  self.featureHighlightView.autoresizingMask =
+      UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
```
This would expect autoresizingMask to be explicitly set / handled in featureHighlightView.
A view controller should only be changing the state of it's child views that it needs to, typically frame.  (And in our case properties that we need to push through, adjustsFont..., colors, fonts)

jverkoey replied:
Autoresizing masks, along with frame, center, and transform are properties that a view should rarely, if ever, set on itself because any view (by definition) lacks the understanding and context of how it will be presented — only the parent has this context.

So in this case it is expected that the owner of the featureHighlightView would set the autoresizingMask to the desired behavior that will allow the controller to maintain the desired layout.

jverkoey replied:
Oh my. Is `featureHighlightView` not a subview?

jverkoey replied:
*facepalm*

Ok. If self.view == self.featureHighlightView then the above statement still holds true, and we definitely should not be setting any of those properties here.

ianegordon replied:
Indeed, self.view == self.featureHighlightView
I think we are in agreement that frame / bounds, center and transform typically include the UIView properties that a view controller is 'allowed' to adjust on its subviews.
I would expect a view to handle it's autoresizingMask and autoresizesSubviews internally.

jverkoey replied:
autoresizingMask is a parent responsibility, autoresizesSubviews is the view's responsibility.

Like .frame (center + bounds + transform), a view lacks the contextual awareness to be able to set an autoresizing mask. A view that enforces its own layout behavior via setting its own autoresizingMask will very quickly break layout in the parent view.

So the concrete requirement for this PR as-is is to remove the autoresizingMask modification because this view controller lacks the parent context required to know if a flexible width/height is truly desired (e.g. maybe the feature highlight view is being shown in a subset of the screen and should have a flexible top/left margin instead — it's the responsibility of the owner of this feature highlight view to decide that).

mohammadcazig replied:
This seems like something we completely need to fix but I actually disagree with making this change here since this PR is unrelated to this topic or issue which already existed in this VC. We can create a separate issue to address that and fix it. I don't have as much context as you two @jverkoey @ianegordon  so would you guys mind creating a github issue for that.

jverkoey replied:
Agreed @mohammadcazig! Filing now.

jverkoey replied:
https://github.com/material-components/material-components-ios/issues/4262

ianegordon replied:
I'm not sure I follow on autoresizingMask.  Why don't we catch up in person tomorrow, with a whiteboard / visual aids?  We are currently not consistent across our components, and we definitely should be.  Thanks for opening the issue.

ianegordon reviewed (approved):
Happy to discuss codifying a pattern offline if that would be useful.

mohammadcazig committed:
moving the string files in to the right folder

mohammadcazig committed:
change the if logic to isViewLoaded for more clarity?

ianegordon commented:
Excellent! Please merge when CI is happy. 🎆
